src/main/java/ch/njol/skript/structures/StructVariables.java

```diff
@@ -99,17 +99,23 @@ public void add(String variable, Class<?>... hints) {
 			if (CollectionUtils.containsAll(hints, Object.class)) // Ignore useless type hint.
 				return;
 			// This important empty check ensures that the variable type hint came from a defined DefaultVariable.
-			if (this.hints.isEmpty())
-				return;
-			this.hints.getFirst().put(variable, hints);
+			synchronized (this.hints) {
```

Comment:
!!! This is not sufficient !!!

Synchronization preserves the happens-before relationship of _operations_ between threads but it will not make this thread-safe by itself!!

Array deques use an internal transient backing array like an array list. Arrays _are_ (to an extent) thread safe (in the sense that all threads should see the same contents rather than using local memory pools) but the backing array object itself and the pointers for the queue are not volatile.
<img width="422" alt="image" src="https://github.com/user-attachments/assets/cb1a52ee-0829-453b-ac6e-c756687e00f2">

This means that there is an edge case where an element is added to the queue on thread 1, exceeding its capacity in either direction, causing it to grow, and while thread 2 does wait its turn, it is not guaranteed to see the correct queue or the correct pointers and may do an incorrect operation, any of:

1. Overwriting an already-filled slot that it sees beyond the head/tail
2. Potentially trying to write beyond the array's capacity (if it sees the change in the head/tail but not the array)
3. Repeating the growth operation and replacing the newly-created array again.

This is of course very unlikely to happen in practice (so was the original error) but it's possible in theory.

The simplest way around this is just to give the ArrayDeque an initial capacity large enough that it's unlikely to have to grow, so this won't happen.
You could also use a different, thread-safe queue/deque implementation.

Reply:
I might prefer to give it a large enough capacity to save on performance. We might be able to make a good estimate of how large it will actually grow

Reply:
The default size is 16 elements. From what I can tell, the queue will only ever change when obtaining the value of a variable. So, to reach 16, I think you would need to have a variable with many variables within it, e.g. `{skript::%{a::%{b::%{c::<and so on>}%}%::*}`. I don't think it's likely that there would ever be more than 16? It might be a good idea to describe this in a comment to show that we are aware of this *potential* issue. What do you think?

Reply:
That's fine by me.

Reply:
Guava has [Queues#synchronizedDeque](https://guava.dev/releases/19.0/api/docs/com/google/common/collect/Queues.html#synchronizedDeque(java.util.Deque))